### PR TITLE
Add support for OS versions in node repo

### PR DESCRIPTION
**Pull request by mpolden**

### CODE REVIEW

**Comment by mpolden** [comment]
1. Controller will set `{"osVersion": "X.Y"}` by PATCHing `/nodes/v2/upgrade`.
2. Nodes will read and act on `wantedOsVersion` from `/nodes/v2/node/<fqdn>`.
3. Nodes will update `currentOsVersion` by PATCHing `/nodes/v2/node/<fqdn>`.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/OsVersions.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+            if (!force && oldTarget.filter(v -> v.isAfter(newTarget)).isPresent()) {
+                throw new IllegalArgumentException("Cannot set target OS version to " + newTarget +
+                                                   " without setting 'force', as it's lower than the current version: "
+                                                   + oldTarget.get());
+            }
+
+            osVersions.put(nodeType, newTarget);
+            db.writeOsVersions(osVersions);
+            currentTargets.set(ImmutableMap.copyOf(osVersions));
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/OsVersions.java]: How will this be updated on other config servers?

**Review by freva — COMMENTED** [review]
Seems scary to trigger OS upgrades for potentially 100+ nodes at the same time with no way to stop it in the middle. Only way to stop is either to set OS version to previous version, thereby forcing to downgrade the nodes that already upgraded, or continue with the upgrade?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/OsVersions.java (hunk 1 above)
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/OsVersions.java]: Good point, will fix.

**Comment by mpolden** [comment]
> Seems scary to trigger OS upgrades for potentially 100+ nodes at the same time with no way to stop it in the middle. Only way to stop is either to set OS version to previous version, thereby forcing to downgrade the nodes that already upgraded, or continue with the upgrade?

Yes, this was discussed during package opening. Roll forward is the only option. Upgrades will roll out per zone and stop if one or more nodes fail to upgrade. Worst case is that we end up with 20% (or whatever the orchestration policy is) of hosts in a zone failing the upgrade.

**Comment by mpolden** [comment]
PTAL

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private final Version version;
+
+    private NodeOsVersionFilter(Version version, NodeFilter next) {
+        super(next);
+        this.version = Objects.requireNonNull(version, "version cannot be null");
+    }
+
+    @Override
+    public boolean matches(Node node) {
+        if (!version.isEmpty() && !node.status().osVersion().filter(v -> v.equals(version)).isPresent()) {
```
Comment (andreer) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java]: I found this statement a bit hard to read, would this be equivalent?
``
        Optional<Version> nodeVersion = node.status().osVersion();
        if (this.version.isEmpty() || !nodeVersion.isPresent()) return nextMatches(node);
        return this.version.equals(nodeVersion.get());
``

Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java]: Thanks, that's better. Will change it.

**Review by andreer — APPROVED** [review]

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java (hunk 2 above)
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java]: Actually, that's not the same thing due to the behavior of `NodeFilter#nextMatches`.

Your suggestion would be fine if `NodeFilter#nextMatches` method didn't return `true` for the last filter in the chain. I think the intention is that filters should match a given node if no filters in the chain actively reject it (a bit strange, but other code depends on this behavior).

**Pull request merged**